openwisp_firmware_upgrader/static/firmware-upgrader/js/batch-upgrade-progress.js

```diff
@@ -0,0 +1,433 @@
+  let imageDisplay = data.image_name || "None";
+  let modifiedTime = data.modified
+    ? getFormattedDateTimeString(data.modified)
+    : "Just now";
+
+  let newRowHtml = `
+    <tr class="${rowClass}">
+      <td>
+        <a href="${deviceUrl}" class="device-link" aria-label="View device ${data.device_name}">
+          ${data.device_name}
+        </a>
+      </td>
+      <td class="status-cell" data-operation-id="${data.operation_id}">
+        <div class="status-content">${data.status}</div>
+      </td>
+      <td>${imageDisplay}</td>
+      <td>${modifiedTime}</td>
+    </tr>
+  `;
+
+  tbody.append(newRowHtml);
+  let newRow = tbody.find(`tr:last`);
+  let statusCell = newRow.find(".status-cell");
+  let operation = {
+    status: data.status,
+    id: data.operation_id,
+    progress: data.progress,
+  };
+  updateBatchStatusWithProgressBar(statusCell, operation);
+}
+
+function updateBatchStatusWithProgressBar(statusCell, operation) {
+  let $ = django.jQuery;
+  let status = operation.status;
+  let progressPercentage = getBatchProgressPercentage(status, operation.progress);
+  statusCell.empty();
+  statusCell.append('<div class="upgrade-status-container"></div>');
+  let statusContainer = statusCell.find(".upgrade-status-container");
+  let statusHtml = "";
+
+  if (status === "in-progress" || status === "in progress") {
+    statusHtml = `<div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill in-progress" style="width: ${progressPercentage}%"></div>
+      </div>`;
+  } else if (status === "success" || status === "completed successfully") {
+    statusHtml = `<div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill success" style="width: 100%"></div>
+      </div>`;
+  } else if (status === "failed") {
+    statusHtml = `<div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill failed" style="width: 100%"></div>
+      </div>`;
+  } else if (status === "aborted") {
+    statusHtml = `<div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill aborted" style="width: 100%"></div>
+      </div>`;
+  } else if (status === "cancelled") {
+    statusHtml = `<div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill cancelled" style="width: 100%"></div>
+      </div>`;
+  } else {
+    statusHtml = `<div class="upgrade-progress-bar">
+        <div class="upgrade-progress-fill" style="width: ${progressPercentage}%"></div>
+      </div>`;
+  }
+  statusContainer.html(statusHtml);
+}
+
+function getBatchProgressPercentage(status, operationProgress = null) {
+  if (operationProgress !== null && operationProgress !== undefined) {
+    return Math.min(100, Math.max(5, operationProgress));
+  }
+  if (
+    status === "completed successfully" ||
+    status === "success" ||
+    status === "failed" ||
+    status === "aborted" ||
+    status === "cancelled"
+  ) {
+    return 100;
+  }
```

Comment:
(same here) I see this pattern of string comparisons frequently in our `.js` files. Can we make this cleaner by using a `const object` (since JS doesn't have native enums)?